File: Suspension matlab/Natural_frequency/Overall_rigidity/Overall_rigidity.m
%% Overall rigidity
clear; clc; close all;
scriptInfo = dbstack('-completenames');
scriptDir = fileparts(scriptInfo(1).file);
addpath(scriptDir);
params = load_data(fullfile(scriptDir,'data.csv'));
paramNames = fieldnames(params);
for paramIdx = 1:numel(paramNames)
    eval([paramNames{paramIdx} ' = params.(paramNames{paramIdx});']);
end

%% 固定參數
Ks = overall_Ks;      % 懸吊剛性 (N/m)
Kt = overall_Kt;      % 輪胎剛性 (N/m)

%% 車體剛性掃描 (1~20倍懸吊剛性)
ratio = overall_ratio;
Kc_list = ratio * Ks;

Keq_list = zeros(size(Kc_list));

%% 計算等效剛性
for i = 1:length(Kc_list)
    Kc = Kc_list(i);
    
    Keq_list(i) = 1 / (1/Kt + 1/Ks + 1/Kc);
end

%% 畫圖
figure;
plot(ratio, Keq_list, 'o-', 'LineWidth', 2);
grid on;

xlabel('Chassis stiffness / Suspension stiffness ratio (Kc / Ks)');
ylabel('Equivalent stiffness Keq (N/m)');
title('Effect of Chassis Stiffness on Overall Stiffness');

%% 標記 10 倍區域
hold on;

xline(10, '--r', '10x threshold', 'LineWidth', 2);

%% 顯示飽和值
Keq_inf = 1 / (1/Kt + 1/Ks);
yline(Keq_inf, '--k', 'Asymptote', 'LineWidth', 2);

legend('Keq', '10x rule', 'Saturation limit');

File: Suspension matlab/Natural_frequency/Overall_rigidity/Overall_rigidity.py
from pathlib import Path
import sys

import matplotlib.pyplot as plt
import numpy as np

ROOT = next(p for p in Path(__file__).resolve().parents if (p / "suspension_utils.py").exists())
sys.path.insert(0, str(ROOT))

from suspension_utils import load_data, show_plots


def main() -> None:
    p = load_data(Path(__file__).with_name("data.csv"))
    ks = p["overall_Ks"]
    kt = p["overall_Kt"]
    ratio = np.asarray(p["overall_ratio"], dtype=float)

    kc_list = ratio * ks
    keq_list = 1 / (1 / kt + 1 / ks + 1 / kc_list)
    keq_inf = 1 / (1 / kt + 1 / ks)

    plt.figure()
    plt.plot(ratio, keq_list, "o-", linewidth=2, label="Keq")
    plt.axvline(10, linestyle="--", color="r", linewidth=2, label="10x rule")
    plt.axhline(keq_inf, linestyle="--", color="k", linewidth=2, label="Saturation limit")
    plt.grid(True)
    plt.xlabel("Chassis stiffness / Suspension stiffness ratio (Kc / Ks)")
    plt.ylabel("Equivalent stiffness Keq (N/m)")
    plt.title("Effect of Chassis Stiffness on Overall Stiffness")
    plt.legend()
    show_plots()


if __name__ == "__main__":
    main()
